crates/control/src/obstacle_filter.rs

```diff
@@ -63,12 +66,15 @@ impl ObstacleFilter {
     pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
         let field_dimensions = context.field_dimensions;
         let cycle_start_time = context.cycle_time.start_time;
-        let measured_robots = context
+        let measurements = context
             .detected_feet_top
             .persistent
             .iter()
-            .zip(context.detected_feet_bottom.persistent.values());
-        for ((detection_time, robots_top), robots_bottom) in measured_robots {
+            .zip(context.detected_feet_bottom.persistent.values())
+            .zip(context.detected_robots_top.persistent.values())
+            .zip(context.detected_robots_bottom.persistent.values());
```

Comment:
You are right, they are all the same length, but not because they come from the same cycler.
They are the same length because the finalization of future queue items happens simultaneously for all perception cycler values.
Each item in this zip iterator is a tuple of timestamp and vecs. The vecs inside can be of different lengths (if they come from different perception cyclers).